org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolServer.java

```diff
@@ -0,0 +1,229 @@
+     *              event content
+     */
+    private void sendEvent(String eventType, Object body) {
+        sendMessage(new Messages.Event(eventType, body));
+    }
+
+    /**
+     * If the the dispatcher is idle, then send the event immediately.
+     * Else add the new event to an eventQueue first and send them when dispatcher becomes idle again.
+     * @param eventType
+     *              event type
+     * @param body 
+     *              event content
+     */
+    private void sendEventLater(String eventType, Object body) {
+        synchronized (this.lock) {
+            if (this.isDispatchingData) {
+                this.eventQueue.offer(new Messages.Event(eventType, body));
+            } else {
+                sendMessage(new Messages.Event(eventType, body));
+            }
+        }
+    }
+
+    private void sendMessage(Messages.ProtocolMessage message) {
+        message.seq = this.sequenceNumber++;
+
+        String jsonMessage = JsonUtils.toJson(message);
+        char[] jsonBytes = jsonMessage.toCharArray();
+
+        String header = String.format("Content-Length: %d%s", jsonBytes.length, TWO_CRLF);
+        char[] headerBytes = header.toCharArray();
+
+        char[] data = new char[headerBytes.length + jsonBytes.length];
+        System.arraycopy(headerBytes, 0, data, 0, headerBytes.length);
+        System.arraycopy(jsonBytes, 0, data, headerBytes.length, jsonBytes.length);
+
+        try {
+            Logger.logInfo("\n[[RESPONSE]]\n" + new String(data));
+            this.writer.write(data, 0, data.length);
+            this.writer.flush();
+        } catch (IOException e) {
+            Logger.logException("Write data to io exception", e);
+        }
+    }
+
+    private void processData() {
+        while (true) {
+            if (this.bodyLength >= 0) {
+                if (this.rawData.length() >= this.bodyLength) {
+                    char[] buf = this.rawData.removeFirst(this.bodyLength);
+                    this.bodyLength = -1;
+                    dispatchRequest(new String(buf));
+                    continue;
+                }
+            } else {
+                String body = this.rawData.getString();
+                int idx = body.indexOf(TWO_CRLF);
+                if (idx != -1) {
+                    Matcher matcher = CONTENT_LENGTH_MATCHER.matcher(body);
+                    if (matcher.find()) {
+                        this.bodyLength = Integer.parseInt(matcher.group(1));
+                        this.rawData.removeFirst(idx + TWO_CRLF.length());
+                        continue;
+                    }
+                }
+            }
+            break;
+        }
+    }
+
+    private void dispatchRequest(String request) {
+        try {
+            Logger.logInfo("\n[REQUEST]\n" + request);
+            Messages.Request message = JsonUtils.fromJson(request, Messages.Request.class);
+            if (message.type.equals("request")) {
+                synchronized (this.lock) {
```

Comment:
Actually the lock object is just used to control the synchronized access to the isDispatchingData variable because java doesn't support invoking synchronized(boolean). Yes, dispatchRequest is only being called in a single thread. But isDispatchingData field could be accessed by two threads (dispatch thread and event thread). Reference these two methods `dispatchRequest` and `sendEventLater`, they could be ran in different threads and access isDispatchingData simultaneously.